File: myExpenses/src/test/java/org/totschnig/myexpenses/provider/ProviderUtilsTest.java
package org.totschnig.myexpenses.provider;

import android.content.OperationApplicationException;
import android.os.Bundle;
import android.os.RemoteException;

import org.junit.After;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.totschnig.myexpenses.model.Account;
import org.totschnig.myexpenses.model.AccountType;
import org.totschnig.myexpenses.model.Category;
import org.totschnig.myexpenses.model.CurrencyUnit;
import org.totschnig.myexpenses.model.Money;
import org.totschnig.myexpenses.model.PaymentMethod;
import org.totschnig.myexpenses.model.Transaction;
import org.totschnig.myexpenses.model.Transfer;

import java.util.Currency;
import java.util.concurrent.TimeUnit;

import static org.totschnig.myexpenses.contract.TransactionsContract.Transactions.ACCOUNT_LABEL;
import static org.totschnig.myexpenses.contract.TransactionsContract.Transactions.AMOUNT_MICROS;
import static org.totschnig.myexpenses.contract.TransactionsContract.Transactions.CATEGORY_LABEL;
import static org.totschnig.myexpenses.contract.TransactionsContract.Transactions.COMMENT;
import static org.totschnig.myexpenses.contract.TransactionsContract.Transactions.CURRENCY;
import static org.totschnig.myexpenses.contract.TransactionsContract.Transactions.DATE;
import static org.totschnig.myexpenses.contract.TransactionsContract.Transactions.METHOD_LABEL;
import static org.totschnig.myexpenses.contract.TransactionsContract.Transactions.OPERATION_TYPE;
import static org.totschnig.myexpenses.contract.TransactionsContract.Transactions.PAYEE_NAME;
import static org.totschnig.myexpenses.contract.TransactionsContract.Transactions.REFERENCE_NUMBER;
import static org.totschnig.myexpenses.contract.TransactionsContract.Transactions.TRANSFER_ACCOUNT_LABEL;
import static org.totschnig.myexpenses.contract.TransactionsContract.Transactions.TYPE_TRANSFER;

@RunWith(RobolectricTestRunner.class)
public class ProviderUtilsTest {
  private Account euroAccount;
  private Account dollarAccount;

  @Before
  public void setupAccounts() {

    euroAccount = new Account("EUR-Account", new CurrencyUnit(Currency.getInstance("EUR")), 0L, null, AccountType.CASH, Account.DEFAULT_COLOR);
    euroAccount.save();
    dollarAccount = new Account("USD-Account", new CurrencyUnit(Currency.getInstance("USD")), 0L, null, AccountType.CASH, Account.DEFAULT_COLOR);
    dollarAccount.save();
  }

  @After
  public void deleteAccounts() throws RemoteException, OperationApplicationException {
    Account.delete(euroAccount.getId());
    Account.delete(dollarAccount.getId());
  }

  @Test
  public void shouldPickAccountBasedOnCurrency() {
    Bundle extras = new Bundle();
    extras.putString(CURRENCY, "EUR");
    Transaction transaction = ProviderUtils.buildFromExtras(extras);
    Assert.assertEquals(euroAccount.getId(), transaction.getAccountId());
    extras = new Bundle();
    extras.putString(CURRENCY, "USD");
    transaction = ProviderUtils.buildFromExtras(extras);
    Assert.assertEquals(dollarAccount.getId(), transaction.getAccountId());
  }

  @Test
  public void shouldPickAccountBasedOnLabel()  {
    Bundle extras = new Bundle();
    extras.putString(ACCOUNT_LABEL, "EUR-Account");
    Transaction transaction = ProviderUtils.buildFromExtras(extras);
    Assert.assertEquals(euroAccount.getId(), transaction.getAccountId());
    extras = new Bundle();
    extras.putString(ACCOUNT_LABEL, "USD-Account");
    transaction = ProviderUtils.buildFromExtras(extras);
    Assert.assertEquals(dollarAccount.getId(), transaction.getAccountId());
  }

  @Test
  public void shouldSetAmount() {
    Bundle extras = new Bundle();
    extras.putString(CURRENCY, "EUR");
    extras.putLong(AMOUNT_MICROS, 1230000);
    Transaction transaction = ProviderUtils.buildFromExtras(extras);
    Assert.assertEquals(new Money(new CurrencyUnit(Currency.getInstance("EUR")), 123L), transaction.getAmount());
  }

  @Test
  public void shouldSetDate() {
    Bundle extras = new Bundle();
    long date = (System.currentTimeMillis() + TimeUnit.DAYS.toMillis(2)) / 1000;
    extras.putLong(DATE, date);
    Transaction transaction = ProviderUtils.buildFromExtras(extras);
    Assert.assertEquals(date, transaction.getDate());
  }

  @Test
  public void shouldSetPayee() {
    Bundle extras = new Bundle();
    String payee = "John Doe";
    extras.putString(PAYEE_NAME, payee);
    Transaction transaction = ProviderUtils.buildFromExtras(extras);
    Assert.assertEquals(payee, transaction.getPayee());
  }

  @Test
  public void shouldSetMainCategory() {
    Bundle extras = new Bundle();
    String category = "A";
    extras.putString(CATEGORY_LABEL, category);
    Transaction transaction = ProviderUtils.buildFromExtras(extras);
    Assert.assertEquals(Category.find(category, null), transaction.getCatId().longValue());
    Assert.assertEquals(category, transaction.getLabel());
  }

  @Test
  public void shouldSetSubCategory() {
    Bundle extras = new Bundle();
    String category = "B:C";
    extras.putString(CATEGORY_LABEL, category);
    Transaction transaction = ProviderUtils.buildFromExtras(extras);
    Assert.assertEquals(Category.find("C", Category.find("B", null)), transaction.getCatId().longValue());
    Assert.assertEquals(category, transaction.getLabel());
  }

  @Test
  public void shouldSetComment() {
    Bundle extras = new Bundle();
    String comment = "A note";
    extras.putString(COMMENT, comment);
    Transaction transaction = ProviderUtils.buildFromExtras(extras);
    Assert.assertEquals(comment, transaction.getComment());
  }

  @Test
  public void shouldSetMethod() {
    Bundle extras = new Bundle();
    String method = "CHEQUE";
    extras.putString(METHOD_LABEL, method);
    Transaction transaction = ProviderUtils.buildFromExtras(extras);
    Assert.assertEquals(PaymentMethod.find(method), transaction.getMethodId().longValue());
  }

  @Test
  public void shouldSetReferenceNumber() {
    Bundle extras = new Bundle();
    String number = "1";
    extras.putString(REFERENCE_NUMBER, number);
    Transaction transaction = ProviderUtils.buildFromExtras(extras);
    Assert.assertEquals(number, transaction.getReferenceNumber());
  }
  @Test
  public void shouldBuildTransfer() {
    Bundle extras = new Bundle();
    extras.putInt(OPERATION_TYPE, TYPE_TRANSFER);
    extras.putString(ACCOUNT_LABEL, "EUR-Account");
    extras.putString(TRANSFER_ACCOUNT_LABEL, "USD-Account");
    Transaction transaction = ProviderUtils.buildFromExtras(extras);
    Assert.assertTrue(transaction instanceof Transfer);
    Assert.assertEquals(dollarAccount.getId(), transaction.getTransferAccountId().longValue());
    Assert.assertEquals("USD-Account", transaction.getLabel());
  }
}

File: myExpenses/src/test/java/org/totschnig/myexpenses/provider/ProviderUtilsTest.kt
package org.totschnig.myexpenses.provider

import android.os.Bundle
import org.junit.Assert
import org.junit.Assert.assertEquals
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.totschnig.myexpenses.BaseTestWithRepository
import org.totschnig.myexpenses.contract.TransactionsContract.Transactions
import org.totschnig.myexpenses.db2.findCategory
import org.totschnig.myexpenses.db2.findPaymentMethod
import org.totschnig.myexpenses.model.Transfer
import java.util.concurrent.TimeUnit

@RunWith(RobolectricTestRunner::class)
class ProviderUtilsTest: BaseTestWithRepository() {

    private var euroAccount: Long = 0
    private var dollarAccount: Long = 0

    @Before
    fun setupAccounts() {

        euroAccount = insertAccount(
            label = "EUR-Account",
            currency = "EUR"
        )
        dollarAccount = insertAccount(
            label = "USD-Account",
            currency = "USD"
        )
    }

    private fun buildFromExtras(extras: Bundle) = ProviderUtils.buildFromExtras(repository, extras)!!

    @Test
    fun shouldPickAccountBasedOnCurrency() {
        var extras = Bundle()
        extras.putString(Transactions.CURRENCY, "EUR")
        var transaction = buildFromExtras(extras)
        assertEquals(euroAccount, transaction.accountId)
        extras = Bundle()
        extras.putString(Transactions.CURRENCY, "USD")
        transaction = buildFromExtras(extras)
        assertEquals(dollarAccount, transaction.accountId)
    }

    @Test
    fun shouldPickAccountBasedOnLabel() {
        var extras = Bundle()
        extras.putString(Transactions.ACCOUNT_LABEL, "EUR-Account")
        var transaction = buildFromExtras(extras)
        assertEquals(euroAccount, transaction.accountId)
        extras = Bundle()
        extras.putString(Transactions.ACCOUNT_LABEL, "USD-Account")
        transaction = buildFromExtras(extras)
        assertEquals(dollarAccount, transaction.accountId)
    }

    @Test
    fun shouldSetAmount() {
        val extras = Bundle()
        extras.putString(Transactions.CURRENCY, "EUR")
        extras.putLong(Transactions.AMOUNT_MICROS, 1230000)
        val transaction = buildFromExtras(extras)
        assertEquals(
            123,
            transaction.amount.amountMinor
        )
    }

    @Test
    fun shouldSetDate() {
        val extras = Bundle()
        val date = (System.currentTimeMillis() + TimeUnit.DAYS.toMillis(2)) / 1000
        extras.putLong(Transactions.DATE, date)
        val transaction = buildFromExtras(extras)
        assertEquals(date, transaction.date)
    }

    @Test
    fun shouldSetPayee() {
        val extras = Bundle()
        val payee = "John Doe"
        extras.putString(Transactions.PAYEE_NAME, payee)
        val transaction = buildFromExtras(extras)
        assertEquals(payee, transaction.party?.name)
    }

    @Test
    fun shouldSetMainCategory() {
        val extras = Bundle()
        val category = "A"
        extras.putString(Transactions.CATEGORY_LABEL, category)
        val transaction = buildFromExtras(extras)
        assertEquals(repository.findCategory(category), transaction.catId)
        assertEquals(category, transaction.categoryPath)
    }

    @Test
    fun shouldSetSubCategory() {
        val extras = Bundle()
        val category = "B:C"
        extras.putString(Transactions.CATEGORY_LABEL, category)
        val transaction = buildFromExtras(extras)
        assertEquals(
            repository.findCategory("C", repository.findCategory("B")),
            transaction.catId
        )
        assertEquals(category, transaction.categoryPath)
    }

    @Test
    fun shouldSetComment() {
        val extras = Bundle()
        val comment = "A note"
        extras.putString(Transactions.COMMENT, comment)
        val transaction = buildFromExtras(extras)
        assertEquals(comment, transaction.comment)
    }

    @Test
    fun shouldSetMethod() {
        val extras = Bundle()
        val method = "CHEQUE"
        extras.putString(Transactions.METHOD_LABEL, method)
        val transaction = buildFromExtras(extras)
        assertEquals(repository.findPaymentMethod(method), transaction.methodId)
    }

    @Test
    fun shouldSetReferenceNumber() {
        val extras = Bundle()
        val number = "1"
        extras.putString(Transactions.REFERENCE_NUMBER, number)
        val transaction = buildFromExtras(extras)
        assertEquals(number, transaction.referenceNumber)
    }

    @Test
    fun shouldBuildTransfer() {
        val extras = Bundle()
        extras.putInt(Transactions.OPERATION_TYPE, Transactions.TYPE_TRANSFER)
        extras.putString(Transactions.ACCOUNT_LABEL, "EUR-Account")
        extras.putString(Transactions.TRANSFER_ACCOUNT_LABEL, "USD-Account")
        val transaction = buildFromExtras(extras)
        Assert.assertTrue(transaction is Transfer)
        assertEquals(dollarAccount, transaction.transferAccountId.toLong())
    }
}